Generate a context if one does not exist

jimhester opened the pull request:
Test files without an explicit `context()` would cause

    Error in x[[method]](...) : attempt to apply non-function

errors in the Progress reporter. This generates a
context from the filename if there is not already a defined context,
fixing the issue.

Fixes #700
Fixes #705

gaborcsardi commented:
Should we cut the `test-` prefix?

jimhester commented:
I am not really sure how we should test this. I was using a test file like the following

```r
test_that("multiplication works", {
  expect_equal(2 * 2, 4)
})

context("blah")
context("blah2")

test_that("multiplication works", {
  expect_equal(2 * 2, 4)
})

context("baz")

test_that("multiplication works", {
  expect_equal(2 * 2, 4)
})

context("qux")
```

Which should generate output like

```
✔ | OK F W S | Context
✔ |  1       | test-blah.R
✔ |  0       | blah
✔ |  1       | blah2
✔ |  1       | baz
✔ |  0       | qux

══ Results ══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
OK:       3
Failed:   0
Warnings: 0
Skipped:  0
```

jimhester commented:
I wasn't sure about cutting the prefix (and possibly the extension?) or not. On one hand they are mostly redundant information, but on the other having the filename makes it easy to navigate to the relevant test file.

Keeping the full filename also makes it easy to transition to removing contexts #728, you can just remove the `context()` calls from the test files.

gaborcsardi commented:
Looks like a good test to me.  But maybe I would compare the returned result of the reporter, not a the output? Or turn off Unicode and colors in the output.

jimhester commented on R/reporter.R:
```diff
@@ -13,10 +13,12 @@ Reporter <- R6::R6Class("Reporter",
     end_test =       function(context, test) {},
     end_context =    function(context) {},
     end_reporter =   function() {},
+    end_file =       function() {},
```
I added `end_file` even though I don't actually use it because we may need it in the future if we are removing context in favor of filenames.

jimhester commented on R/test-files.R:
```diff
@@ -126,6 +126,7 @@ test_file <- function(path, reporter = default_reporter(), env = test_env(),
     reporter = reporter,
     start_end_reporter = start_end_reporter,
     {
+      get_reporter()$start_file(basename(path))
```
We need to do this separately from the lister, because the lister is not part of the default reporter.

gaborcsardi replied:
this would make a good comment in the source imo

gaborcsardi commented:
These are good points re cutting, but I think if we are to deprecate `context()` then it makes sense to still cut it, because it will be just a lot of redundant info.

Usually you want to go to the test file if sg fails there, and then you'll have the file name in the failure.

If we don't deprecate `context()` then it makes sense not to cut.

jimhester commented on R/reporter-progress.R:
```diff
@@ -57,6 +58,16 @@ ProgressReporter <- R6::R6Class("ProgressReporter",
+
+    start_test = function(context, test) {
+      if (is.null(context)) {
+        context(self$file_name)
```
Using the regular `context()` function rather than `self$start_context()` or `self$.start_context()` because for the multi-reporter we need to assign `self$.context` in the multi-reporter rather than in the contained reporters.

gaborcsardi replied:
this one as well

gaborcsardi commented on NEWS.md:
```diff
@@ -8,6 +8,8 @@
 
 * Pass through warnings when `options(warn = 2)` is set (#721, @yutannihilation).
+* Progress reporter now generates a context from the filename and no longer
+  errors if tests occur before a context (#700, #705).
```
Other reporters are OK without a context?

jimhester replied:
Yeah other than the JUnit one the rest seem ok

jimhester commented:
Ok I added a few additional contexts to the `tests.R` file, which picked up a similar bug in the JUnit reporter when there is no context (also fixed). The rest of the reporters seem ok, but the expected outputs all changed as a result, so the diff is now big.

hadley commented:
Yeah, I think we should remove the `test-` prefix and the `.R` suffix.  We'll also need to tweak the default context that usethis generates.

hadley commented on R/reporter-progress.R:
```diff
@@ -57,6 +58,22 @@ ProgressReporter <- R6::R6Class("ProgressReporter",
+      }
+    },
+
     start_context = function(context) {
```
Should this all be moved to `start_file` or will you ignore `context()` calls in a different PR?

jimhester replied:
I was going to do that / discuss it in a separate PR

jimhester commented:
Thank you both for the reviews!
